### PR TITLE
Generate serialization for SecKeychainItemRef

### PR DESCRIPTION
**Pull request by mscottapple**
#### edf8486e0ced3bf3f416bd255aa8b694d7ba01c5
<pre>
Generate serialization for SecKeychainItemRef
<a href="https://bugs.webkit.org/show_bug.cgi?id=268182">https://bugs.webkit.org/show_bug.cgi?id=268182</a>
<a href="https://rdar.apple.com/121676431">rdar://121676431</a>

Reviewed by achristensen07 (Alex Christensen).

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder&lt;SecKeychainItemRef&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RetainPtr&lt;SecKeychainItemRef&gt;&gt;::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h: Added.
(WebKit::CoreIPCSecKeychainItem::CoreIPCSecKeychainItem):
(WebKit::CoreIPCSecKeychainItem::createSecKeychainItem const):
(WebKit::CoreIPCSecKeychainItem::dataReference const):
(WebKit::CoreIPCSecKeychainItem::persistentRefForKeychainItem const):
* Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(getTempKeychain):
(destroyTempKeychain):
(TEST):

Canonical link: <a href="https://commits.webkit.org/274043@main">https://commits.webkit.org/274043@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2bc7b6145342ab109e1ce8b1dbe24096dddd98fb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37631 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39877 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40160 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33499 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38948 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13670 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31878 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38196 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13956 "Found 1 new test failure: media/video-ended-does-not-hold-sleep-assertion.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32975 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12132 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12126 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33657 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41421 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34078 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34081 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37973 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10196 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36142 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14098 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13062 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4887 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13411 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    CoreIPCSecKeychainItem(const IPC::DataReference& data)
+        : m_persistentRef(adoptCF(CFDataCreate(kCFAllocatorDefault, data.data(), data.size())))
+    {
+    }
+
+    RetainPtr<SecKeychainItemRef> createSecKeychainItem() const
+    {
+        RELEASE_ASSERT(hasProcessPrivilege(ProcessPrivilege::CanAccessCredentials));
```
Comment (mscottapple) [inline Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h:58]: NB: This was a `RELEASE_ASSERT` in the original code.

File: Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+        SecKeychainItemRef keychainItem = NULL;
+        CFDataRef data = m_persistentRef.get();
+        ASSERT(data);
+
+        // SecKeychainItemCopyFromPersistentReference() cannot handle 0-length CFDataRefs.
+        if (CFDataGetLength(data)) {
+            ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+            SecKeychainItemCopyFromPersistentReference(data, &keychainItem);
```
Comment (mscottapple) [inline Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h]: There aren't any API/SPI to manually deconstruct a `SecKeychainItem` into component parts. Fortunately, a persistent reference is like an alias or PK to the original item—not the item itself—so there _should_ be less room for mischief with attacker data.

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -3,7 +3,7 @@
 	archiveVersion = 1;
 	classes = {
 	};
-	objectVersion = 54;
+	objectVersion = 55;
```
Comment (achristensen07) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: This change shouldn't be included in this PR.

File: Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm [hunk 4]
```diff
@@ -412,6 +415,35 @@ String certDataBase64(""
...
 
+#if HAVE(SEC_KEYCHAIN)
+static SecKeychainRef getTempKeychain()
+{
+    SecKeychainRef keychainRef = NULL;
+    uuid_t uu;
+    char pass[PATH_MAX];
+
+    uuid_generate_random(uu);
+    uuid_unparse_upper(uu, pass);
```
Comment (achristensen07) [inline Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:513]: pass's length of PATH_MAX is a little strange.  Why not just use NSUUID.UUID and %@?

File: Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm [hunk 5]
```diff
@@ -412,6 +415,35 @@ String certDataBase64(""
...
+    UInt32 passLen = (UInt32)strlen(pass);
+    NSString* path = [NSString stringWithFormat:@"%@/%s", NSTemporaryDirectory(), pass];
+
+    [NSFileManager.defaultManager removeItemAtPath:path error:NULL];
+    RELEASE_ASSERT(SecKeychainCreate(path.fileSystemRepresentation, passLen, pass, false, NULL, &keychainRef) == errSecSuccess);
+    RELEASE_ASSERT(keychainRef);
+    RELEASE_ASSERT(SecKeychainUnlock(keychainRef, passLen, pass, TRUE) == errSecSuccess);
```
Comment (achristensen07) [inline Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm]: Do we need a corresponding SecKeychainLock call?  Also, RELEASE_ASSERT is probably a bit excessive here.  EXPECT_TRUE, EXPECT_NOT_NULL, or EXPECT_EQ cause a test failure without a crash.

File: Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            ASSERT(keychainItem);
+            ALLOW_DEPRECATED_DECLARATIONS_END
+        }
+        return adoptCF(keychainItem);
+    }
+
+    IPC::DataReference dataReference() const
+    {
+        RELEASE_ASSERT(m_persistentRef);
```
Comment (achristensen07) [inline Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h]: We probably don't want to introduce a crash here.  Just returning an empty data reference is a reasonable behavior here.  The hasProcessPrivilege assertions are ok though.

File: Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.serialization.in [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# Copyright (C) 2023 Apple Inc. All rights reserved.
```
Comment (achristensen07) [inline Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.serialization.in]: 2024

File: Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm [hunk 8]
```diff
@@ -644,6 +676,29 @@ String certDataBase64(""
...
+    // SecKeychainItem
+#if HAVE(SEC_KEYCHAIN)
+    ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    // Store the certificate created above into the temp keychain
+    SecKeychainRef tempKeychain = getTempKeychain();
+    CFDataRef certData = NULL;
+    RELEASE_ASSERT(SecItemExport(cert.get(), kSecFormatX509Cert, kSecItemPemArmour, nil, &certData) == errSecSuccess);
+    CFArrayRef items = NULL;
+    RELEASE_ASSERT(SecKeychainItemImport(certData, CFSTR(".pem"), NULL, NULL, 0, NULL, tempKeychain, &items) == errSecSuccess);
+    RELEASE_ASSERT(items);
```
Comment (achristensen07) [inline Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm]: EXPECT_NOT_NULL.

**Review by achristensen07 — CHANGES_REQUESTED** [review]

File: Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm (hunk 4 above)
Comment (mscottapple) [reply Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:513]: Because I also use `pass` a few lines later as the C-string password for the keychain. (This is code from a previous life which I basically copied/pasted.) I could use a value smaller than `PATH_MAX` to be sure, but that's been the historical value to use when doing anything path-related.)

File: Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm (hunk 5 above)
Comment (mscottapple) [reply Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm]: There's no need to lock this keychain, it gets destroyed instead. This is just a throw-away keychain for API which needs a keychain to operate correctly.

I will change the macros.

**Review by achristensen07 — APPROVED** [review]

File: Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        CFDataRef data = m_persistentRef.get();
+        // SecKeychainItemCopyFromPersistentReference() cannot handle 0-length CFDataRefs.
+        if (!CFDataGetLength(data))
+            return nullptr;
+
+        ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+        SecKeychainItemRef keychainItem = NULL;
+        SecKeychainItemCopyFromPersistentReference(data, &keychainItem);
+        ALLOW_DEPRECATED_DECLARATIONS_END
+        return (keychainItem ? adoptCF(keychainItem) : nullptr);
```
Comment (achristensen07) [inline Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h]: This is an unnecessary null check.  adoptCF can be called on null.

**Review by achristensen07 — APPROVED** [review]
